pkg/lib/serverTime.js

```diff
@@ -321,13 +324,181 @@ export function ServerTime() {
                     return result;
                 });
-    };
+    }
 
-    self.set_custom_ntp = function (servers, enabled) {
-        const text = `# This file is automatically generated by Cockpit\n\n[Time]\n${enabled ? "" : "#"}NTP=${servers.join(" ")}\n`;
+    function set_ntp_config_timesyncd(config) {
+        const custom_ntp_config_file = cockpit.file("/etc/systemd/timesyncd.conf.d/50-cockpit.conf",
+                                                    { superuser: "try" });
+
+        const text = `# This file is automatically generated by Cockpit\n\n[Time]\n${config.enabled ? "" : "#"}NTP=${config.servers.join(" ")}\n`;
 
         return cockpit.spawn(["mkdir", "-p", "/etc/systemd/timesyncd.conf.d"], { superuser: "try" })
                 .then(() => custom_ntp_config_file.replace(text));
+    }
+
+    function find_file_and_read(names, first) {
+        if (first === undefined)
+            first = 0;
+        if (first >= names.length)
+            return Promise.reject(new Error("Found none of " + names.join(",")));
+        return cockpit.file(names[first], { superuser: "try" }).read()
+                .then(text => {
+                    if (text)
+                        return [names[first], text];
+                    return find_file_and_read(names, first + 1);
+                });
+    }
+
+    function get_ntp_config_chronyd() {
+        return find_file_and_read(["/etc/chrony.conf", "/etc/chrony/chrony.conf"])
+                .then(([name, text]) => {
+                    const result = {
+                        supported: true,
+                        backend: "chronyd",
+                        file: name,
+                        standard: [],
+                        sources: [],
+                        extra: []
+                    };
+
+                    let in_cockpit_section = false;
+                    for (const line of text.split("\n")) {
+                        if (line == "# Added by Cockpit") {
+                            in_cockpit_section = true;
+                        } else {
+                            const fields = line.split(" ");
+                            if (fields[0] == "server") {
+                                if (in_cockpit_section) {
+                                    result.extra.push(fields[1]);
+                                } else {
+                                    result.standard.push({ name: fields[1], enabled: true });
+                                }
+                            } else {
+                                in_cockpit_section = false;
+                                if (fields[0] == "pool")
+                                    result.standard.push({ name: fields[1], enabled: true });
+                                if (fields[0] == "#pool" || fields[0] == "#server")
+                                    result.standard.push({ name: fields[1], enabled: false });
+                                if (fields[0] == "sourcedir")
```

Comment:
It feels rather odd to me to treat a whole "sourcedir" as an opaque thing, especially as in practice it's always gonna be just one directory anyway (especially with Debian's standard `sources.d/`). Reading the file names there feels a little more practical, but in the end this is all just guesswork as chronyd doesn't have such a "grouping" concept. If you just track in which file a `server` came from, so that you can do a deletion (or option editing in the future) in the correct file, that feels much simpler and robust.

Reply:
> It feels rather odd to me to treat a whole "sourcedir" as an opaque thing, 

I think it is natural.  We should not try to modify the files in a sourcedir, as they are likely owned by someone else.  For example, files in /run/chrony-dhcp will just get recreated on next boot and there is no point in letting the user change them. But it does make sense to disable the whole "NTP servers from DHCP" facility if it causes trouble.

Reply:
Right, that's a good point -- cockpit shuold only ever edit things in /etc, not in /run/, /usr, or anywhere else (cf. "track where it came from")